### PR TITLE
Web login

### PR DESCRIPTION
**Pull request by lemieux**
It should cover the API spec for the login.

### CODE REVIEW

File: src/js/controllers/chatController.js [hunk 1]
```diff
@@ -352,14 +353,27 @@ module.exports = ViewController.extend({
 
     _getLatestReadTime: function() {
         if (!this.latestReadTs) {
-            this.latestReadTs = parseInt(cookie.parse(document.cookie).sk_latestts || 0);
+            var key = SK_LATEST_TS + '_' + endpoint.appUserId;
```
Comment (mspensieri) [inline src/js/controllers/chatController.js]: Read state should be tracked by `userId`. `appUserId` can pretty much change at any time

File: src/js/main.js [hunk 2]
```diff
@@ -63,131 +75,128 @@ var SupportKit = Marionette.Object.extend({
...
-        });
+            return this.login(options.userId, options.jwt);
+    },
 
-        if (typeof options === 'object') {
-            endpoint.appToken = options.appToken;
-            endpoint.jwt = options.jwt;
-            if (options.serviceUrl) {
-                endpoint.rootUrl = options.serviceUrl;
+    login: function(userId, jwt) {
```
Comment (mspensieri) [inline src/js/main.js]: When a user transitions from no userId to having a userId, we should clear the last read timestamp for the anonymous user, since they will be upgraded into the new `userId` and the conversation will be cleared

File: test/mocks/supportKitServer.mock.js [hunk 3]
```diff
@@ -21,9 +21,22 @@ module.exports = BaseServerMock.extend({
...
+            'POST', /\/api\/appboot/, function(xhr) {
+                var body = JSON.parse(xhr.requestBody);
+                var data = _.extend({}, appData);
+                if (body.userId) {
+                    _.extend(data, {
+                        appUserId: body.userId
+                    });
+
+                    _.extend(data.appUser, {
+                        _id: body.userId
```
Comment (mspensieri) [inline test/mocks/supportKitServer.mock.js]: What's the intent here? `userId` is not analogous to `appUser._id`

File: src/js/controllers/chatController.js (hunk 1 above)
Comment (lemieux) [inline src/js/controllers/chatController.js]: Ok I guess I got mixed up between userId and appUserId

File: src/js/main.js (hunk 2 above)
Comment (lemieux) [inline src/js/main.js]: How would I know it's transitionning from an anonymous or a proper using? both have ids... I guess I could put a flag on the user when they log in without a userID

File: test/mocks/supportKitServer.mock.js (hunk 3 above)
Comment (lemieux) [inline test/mocks/supportKitServer.mock.js]: Mix up

File: src/js/main.js (hunk 2 above)
Comment (lemieux) [inline src/js/main.js]: I guess one will have a userId and the other will not... again.. mixup with appUserId :(

File: src/js/controllers/chatController.js (hunk 1 above)
Comment (lemieux) [inline src/js/controllers/chatController.js]: fixed.

File: src/js/main.js (hunk 2 above)
Comment (lemieux) [inline src/js/main.js]: fixed.

File: test/mocks/supportKitServer.mock.js (hunk 3 above)
Comment (lemieux) [inline test/mocks/supportKitServer.mock.js]: fixed.

File: src/js/controllers/chatController.js [hunk 4]
```diff
@@ -21,6 +23,8 @@ var EmailNotificationView = require('../views/emailNotificationView');
 var ChatInputController = require('../controllers/chatInputController');
 var SettingsController = require('../controllers/settingsController');
 
+var SK_LATEST_TS = 'sk_latestts';
```
Comment (alavers) [inline src/js/controllers/chatController.js]: '`sk_latest_ts'`? </ocd>

File: src/js/controllers/chatController.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
+/* global Promise:false */
```
Comment (alavers) [inline src/js/controllers/chatController.js]: Why are these necessary?

File: src/js/faye.js [hunk 6]
```diff
@@ -3,33 +3,26 @@
...
-    // that only has `then(resolve, reject)` and `all` methods.
-    var deferred = $.Deferred();
-
-    faye.subscribe('/conversations/' + conversationId, function(message) {
+    return faye.subscribe('/conversations/' + conversationId, function(message) {
         vent.trigger('receive:message', message);
-    }).then(function() {
-        deferred.resolve(faye);
-    }, function(err) {
-        console.error('Faye subscription error:', err && err.message);
```
Comment (alavers) [inline src/js/faye.js]: Why remove this error message wrapper?

File: src/js/main.js [hunk 7]
```diff
@@ -51,8 +59,12 @@ var SupportKit = Marionette.Object.extend({
 
     initialize: function() {
```
Comment (alavers) [inline src/js/main.js]: The same API has both init and initialize ಠ_ಠ

I'd much prefer to have our SupportKit global stay "clean" and expose only public API, without our internal _foo stuff and Marionette.Object stuff.

Eg, have this guy be `var Main = Marionette.Object.extend({`

and add a separate supportkit.js wrapping `Main` that exposes `init` `login` etc..

File: src/js/controllers/chatController.js (hunk 5 above)
Comment (lemieux) [inline src/js/controllers/chatController.js]: I don't make the rules, jshint is :p ... I'll move that to the jshint config instead.

File: src/js/controllers/chatController.js (hunk 4 above)
Comment (lemieux) [inline src/js/controllers/chatController.js]: good point

File: src/js/main.js [hunk 8]
```diff
@@ -77,117 +88,122 @@ var SupportKit = Marionette.Object.extend({
...
+                platform: 'web'
+            }
+        };
 
-                var RuleCollection = Backbone.Collection.extend({
-                    model: Rule
-                });
+        if (userId) {
+            data.userId = userId;
+            endpoint.userId = userId;
```
Comment (alavers) [inline src/js/main.js]: Shouldn't endpoint.userId be unset if userId is unspecified?

Comment (alavers) [inline src/js/main.js]: Also, should it be reverted to some previous value if appboot call returns 4xx?

Comment (lemieux) [inline src/js/main.js]: it's done in `endpoint.cleanState()`

File: src/js/main.js (hunk 7 above)
Comment (lemieux) [inline src/js/main.js]: fixed.

File: src/js/faye.js (hunk 6 above)
Comment (lemieux) [inline src/js/faye.js]: es6-promise seems to log errors to the console and reject the promise. It felt useless to have this here.

File: src/js/main.js (hunk 8 above)
Comment (lemieux) [inline src/js/main.js]: As for reverting, I felt like the login action would always be logout first, then login. If login fails, log out still happened.

File: src/js/controllers/chatController.js (hunk 5 above)
Comment (lemieux) [inline src/js/controllers/chatController.js]: fixed

**Comment by alavers** [comment]
:+1:
